File: 0875-koko-eating-bananas/0875-koko-eating-bananas.cpp
class Solution {
public:
    bool canEat(vector<int>&piles , int h , int mid){
        long long cnt = 0;
        
        for(int i = 0 ; i < piles.size() ;i++){
            cnt += (piles[i] + mid - 1) / mid;
        }
        return cnt <= h;
    }
    int minEatingSpeed(vector<int>& piles, int h) {
        int n = piles.size();
        int mini = 1 , maxi = *max_element(begin(piles) , end(piles));
        int ans = INT_MAX;
        while(mini <= maxi){
            int mid = (mini + maxi) / 2;
            if(canEat(piles , h , mid)){
                ans = min(ans , mid);
                maxi = mid - 1;
            }else{
                mini = mid + 1;
            }
        }
        return ans;
    }
};

File: 0875-koko-eating-bananas/0875-koko-eating-bananas.java
class Solution {
    boolean canEat(int[] piles, int h , int mid){
        long  n = piles.length;
        long cnt = 0;
        for(int x : piles){
            cnt += (x + mid - 1) / mid;
        }
        return cnt <= h;
    }
    public int minEatingSpeed(int[] piles, int h) {
        int n = piles.length;
        int l =  1 , r = Arrays.stream(piles).max().getAsInt();
        int ans = Integer.MAX_VALUE;
        while(l <= r){
            int mid = (l + r) / 2;
            if(canEat(piles , h , mid)){
                ans = Math.min(ans, mid);
                r = mid - 1;
            }else{
                l = mid + 1;
            }
        }
        return ans;
    }
}